### PR TITLE
OCPEDGE-902: add SNO control plane high cpu usage alert

### PR DESCRIPTION
**Pull request by qJkee**
Create separate SNO alert for control plane high cpu usage. This alert is also aware of workload partitioning and adjusts threshold if workload partitioning is enabled.
I don't track performance profile because after applying new one, node restarts and I can recreate/update alert threshold with new cores amount

### CODE REVIEW

File: pkg/operator/starter.go [hunk 1]
```diff
@@ -213,8 +215,14 @@ func RunOperator(ctx context.Context, controllerContext *controllercmd.Controlle
 		controllerContext.EventRecorder,
 	).
 		WithConditionalResources(bindata.Asset, []string{"assets/alerts/kube-apiserver-slos-extended.yaml"}, notOnSingleReplicaTopology, nil).
+		WithConditionalResources(bindata.Asset, []string{"assets/alerts/cpu-utilization.yaml"}, notOnSingleReplicaTopology, nil).
```
Comment (vrutkovs) [inline pkg/operator/starter.go:218]: Merge into previous line?

**Review by vrutkovs — CHANGES_REQUESTED** [review]
Looks good to me, cc @dgrisonnet for another pair of eyes

File: bindata/assets/alerts/cpu-utilization-sno.yaml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+          annotations:
+            summary: >-
+              CPU utilization across control plane pods is more than 60% of {{ . }} CPU. High CPU usage usually means that something goes wrong.
+            runbook_url: https://github.com/openshift/runbooks/blob/master/alerts/cluster-kube-apiserver-operator/ExtremelyHighIndividualControlPlaneCPU.md
+            description: >-    
+              This level of CPU utlization of an control plane is probably not a problem under most circumstances, but high levels of utilization may indicate
+              problems with cluster or control plane pods. To manage this alert or modify threshold it in case of false positives see the following link:
+              https://docs.openshift.com/container-platform/latest/monitoring/managing-alerts.html
+          expr: |
+            sum(rate(container_cpu_system_seconds_total{namespace=~"openshift-.*",image!=""}[5m])) / {{ . }} * 100 > 60
```
Comment (dgrisonnet) [inline bindata/assets/alerts/cpu-utilization-sno.yaml:20]: I believe that you should use `container_cpu_usage_seconds_total` as the `system` one will only account for time spend in the kernel space

File: bindata/assets/alerts/cpu-utilization-sno.yaml [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+          for: 10m
```
Comment (dgrisonnet) [inline bindata/assets/alerts/cpu-utilization-sno.yaml:21]: For warning alerts, this should be 1h based on https://github.com/openshift/enhancements/blob/799dc3c7b974c35832daea4004b772a3c0102ee1/enhancements/monitoring/alerting-consistency.md#warning-alerts

File: bindata/assets/alerts/cpu-utilization-sno.yaml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+          labels:
+            namespace: openshift-kube-apiserver
+            severity: warning
+        - alert: ExtremelyHighIndividualControlPlaneCPU
```
Comment (dgrisonnet) [inline bindata/assets/alerts/cpu-utilization-sno.yaml:25]: I opened a separate PR to fix the HA alerts, but I am not a big fan of keeping both the warning and critical alerts here, as I detailed in https://github.com/openshift/cluster-kube-apiserver-operator/pull/1555. IMO we should just keep the critical one around

File: pkg/operator/workloadpartitioning/cpuPartitioningAlertRenderer.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	"github.com/openshift/library-go/pkg/operator/resource/resourceread"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/utils/cpuset"
+)
+
+const defaultCoresNum = 4
```
Comment (dgrisonnet) [inline pkg/operator/workloadpartitioning/cpuPartitioningAlertRenderer.go:22]: Is it really safe to assume that all the SNO clusters that are running without CPU partitioning will have 4 cores? If we ever have a case where the machine has 2,6,8,.. cores, the alerts will become incorrect.

Maybe we could use `runtime.NumCPU()` from the go standard library to get an accurate value for the number of CPU available?

File: bindata/assets/alerts/cpu-utilization-sno.yaml [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+              causing even more CPU pressure.
+              It can also cause failing liveness probes due to slow etcd responsiveness on the backend.
+              If one kube-apiserver fails under this condition, chances are you will experience a cascade as the remaining
+              kube-apiservers are also under-provisioned.
+              To fix this, increase the CPU and memory on your control plane nodes.
+              To manage this alert or modify threshold it in case of false positives see the following link: 
+              https://docs.openshift.com/container-platform/latest/monitoring/managing-alerts.html
+          expr: |
+            sum(rate(container_cpu_usage_seconds_total{namespace=~"openshift-.*",image!=""}[5m])) / {{ . }} * 100 > 90
+          for: 1h
```
Comment (dgrisonnet) [inline bindata/assets/alerts/cpu-utilization-sno.yaml:44]: this is critical severity so we should have this down to like 10 minutes

File: bindata/assets/alerts/cpu-utilization-sno.yaml (hunk 3 above)
Comment (dgrisonnet) [reply bindata/assets/alerts/cpu-utilization-sno.yaml:21]: I am still seeing a `for` clause set to 10 minutes when it should be 1h for warning alerts

File: pkg/operator/workloadpartitioning/cpuPartitioningAlertRenderer.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+		if _, ok := pf.Spec.NodeSelector["node-role.kubernetes.io/master"]; !ok {
+			continue
+		}
+		if pf.Spec.CPU == nil || pf.Spec.CPU.Reserved == nil {
+			continue
+		}
+		return coresInCPUSet(*pf.Spec.CPU.Reserved)
+	}
+
+	return defaultCoresNum, nil
```
Comment (dgrisonnet) [inline pkg/operator/workloadpartitioning/cpuPartitioningAlertRenderer.go:108]: We might need to create a watcher logic that would update the alert whenever the workload partitioning values are updated no?

File: pkg/operator/starter.go [hunk 8]
```diff
@@ -213,8 +215,14 @@ func RunOperator(ctx context.Context, controllerContext *controllercmd.Controlle
...
 		AddKubeInformers(kubeInformersForNamespaces)
 
+	err = workloadpartitioning.CreateSNOAlert(ctx, dynamicClient, infrastructure.Status, controllerContext.EventRecorder)
```
Comment (dgrisonnet) [inline pkg/operator/starter.go:221]: We might need to create a watcher logic that would update the alert whenever the workload partitioning values are updated no?

**Review by dgrisonnet — APPROVED** [review]
/lgtm
/approve

**Pull request merged**